File: runtime/ruby/units/rt/wasm_exception.rb
# A wasm exception in flight; the object itself is the exnref value
# (ADR-19). Deliberately not a Trap subclass: traps must never be caught
# by try_table.
class WasmException < StandardError
  attr_reader :tag, :values

  def initialize(tag, values)
    @tag = tag
    @values = values
    super("wasm exception")
  end
end

File: runtime/perl/units/rt/wasm_exception.pl
# A wasm exception in flight; the object itself is the exnref value.
# Deliberately unrelated to Trap, Exit, and LinkError: try_table's classification step tests for this class alone, so traps and the exit path structurally cannot be caught by catch_all.
sub wasm_exception {
    my ($tag, $values) = @_;
    die bless({ tag => $tag, values => $values }, 'Rt::WasmException');
}
